hazelcast/src/main/java/com/hazelcast/map/impl/recordstore/DefaultRecordStore.java

```diff
@@ -284,17 +284,22 @@ public void forEach(BiConsumer<Data, Record> consumer,
                         boolean backup, boolean includeExpiredRecords) {
 
         long now = getNow();
-        Iterator<Map.Entry<Data, Record>> entries = storage.mutationTolerantIterator();
-        while (entries.hasNext()) {
-            Map.Entry<Data, Record> entry = entries.next();
+        storage.beforeOperation();
```

Comment:
We need to make sure threads are registered only as long as they're making progress so in a relatively short time they refresh their view of epochs, otherwise the idle threads would prevent advancing safe epoch, which is a problem for tstore.

@petrpleshachkov it's done here because EPs can be offloaded to a non-partition thread?

Reply:
In cases like `writeRecordStore` we do call it without enclosing `MapOperation` or any other `Operation`.